### ISSUE TITLE
"Tidy.exe ist keine zulässige Win32-Anwendung" = tidy.exe is not a valid Win32 application

### ISSUE THREAD

**Issue by LWillms** [htacg/tidy-html5#566]
That's the message which I get when I try to execute Tidy 5.4 on my Windows Server 2003. 

I downloaded and put it in place back in March, but I can't remember if I had already tried to use it back then. 

It is on this server where I keep and edit my HTML sources, and it is on this server that Tidy should run.

**Comment by geoffmcl** [comment]
@LWillms thanks for the issue, and **sad** to hear that...

Reading around here and there this can have several possible causes...

1. File is corrupt, bad, or missing.

But assume you have checked this... tried another download... verify the file exists... etc...

2. File is not designed for your version of Windows:

One of the most obvious is you are trying to run a 64-bit Tidy on a 32-bit OS...

But it might also be that certain MSVC runtimes are missing, and need to be installed... but I think that should be a different error message about **missing** DLLS.

Tidy Version 5.4.0 was built with MSVC14 (2015)... the earlier with MSVC10 (2010)... See MS sites [like](https://www.microsoft.com/en-us/download/details.aspx?id=48145), and many others...

And read issue #515, where we were able to supply a `/MT` version... the default is `/MD`... but this was a **LINK** problem, not running the Tidy EXE...

So have you checked the bit-ness of your OS matches the Tidy binary you downloaded? 

Where, and exactly **which** tidy binary **did** you download and install?

3. Blocked by Anti-virus program.

Seems unlikely, but have you scanned the tidy.exe file, or tried with the anit-virus off?

4. Or **something else**... there seems a **lot** of posts about this...

I hope the above offers some ideas... please **advise** more... thanks...

**Comment by LWillms** [comment]
Yes, I have re-downloaded the archive, I have compared them, compared with the 64 bit version. I also tried it on the laptop with still works with WinXP, with the same negative result. I had also looked at #514 and tried tidy-5.5.10-win32-2.zip and tidy-5.5.10-win32-mt.zip -- negative. 

I just tried the 32 bit version on my Win7 machine, and got told that it would not run since a VCruntime140.dll is missing on the PC. 

VCruntime140.dll is included in tidy-5.5.10-win32-2.zip...

Is that compiled with the library for Win10? I had quite some problems with a Silverfast 8.8 scanning program which did not run on my 64bit-win7 PC because the automatic update machinery of MS had stalled for months. But finally got it.

**Comment by LWillms** [comment]
Googling for "VCruntime140.dll" I found an exchange on an HP forum (in German language) which said that the VC++ 2015 Studio runtime had a problem, and that one should download and re-install the Visual C++ Redistributable for Visual Studio 2015 from https://www.microsoft.com/en-US/download/details.aspx?id=48145 

From there: 

_Supported Operating System_

Windows 10 , Windows 7 Service Pack 1, Windows 8, Windows 8.1, Windows Server 2003 Service Pack 2, Windows Server 2008 R2 SP1, Windows Server 2008 Service Pack 2, Windows Server 2012, Windows Vista Service Pack 2, Windows XP Service Pack 3

    For additional information on Operating System support, see the Visual Studio 2015 Compatibility page. 

_Hardware Requirements:_
        1.6 GHz or faster processor
        1 GB of RAM (1.5 GB if running on a virtual machine)
        50 MB of available hard disk space
        5400 RPM hard drive


I downloaded, installed, and unfortunately found no change on the Server 2003.

**Comment by geoffmcl** [comment]
@LWillms, sorry but I am getting a little **lost** ;=()

Is your current Server 2003 32 or 64 bit? You talk about **upgrading** to a `64 bit` Server 2008 R2...

Of course if it is currently 32-bit it can **not** run a 64-bit executable... Usually the 32-bit EXE will run in both OSES...

Then you say `the 5.2 version still does its work`! Is that the 32 or 64 bit? Or both?

Maybe I could try building a MSVC10 version, but do you want a 32-bit or a 64-bit version? I would like to understand and solve this, **if possible**...

Sorry for being so dense... please advise **more**, and keep it real simple for me... thanks...

**Comment by LWillms** [comment]
After having installed the VC 2105 Runtime in both 64 and 32 (x86) bit versions on my 64 bit Win7 computer, the 32-bit Tidy starts to run there without problems. The VCruntime140.dll is being installed with the x86-version of the runtime.

**Comment by LWillms** [comment]
Clarification on my systems: the server is currently 32 bit hardware with the 32 bit server operating system Server 2003. I plan to replace this with a 64 bit machine and run Server 2008 R2 on it -- this OS is available as 64 bit version only. And I have the concrete need to upgrade the server OS because at least one application does no longer run on an XP-level OS. 

But as long as this is not accomplished, I would like to use the 32 bit version of Tidy on the server. Since Tidy 5.2 is still working in that environment, I can renounce on using 5.4. But others with a similar operating system might want to have it. 

Let's see what comes first: Tidy 5.5 or my server on 64 bit hardware and software?

**Comment by geoffmcl** [comment]
@LWillms thanks for the further information... ok, we are talking about **two** systems...

 1. Win7 - 64-bit
 2. Server 2003 - 32-bit
 
So I have some questions...
 
 1. Win7 - 64-bit
 
You seem to be reporting `32-bit Tidy 5.4` runs there, now that you have installed the VCruntime140.dll. That's good... no problem... understood...

But if you also installed the VC 64 (x64) runtimes, then so should `64-bit Tidy 5.4`!

Are you saying this `64-bit Tidy 5.4` **fails**? In Win7 - 64-bit! With what error message?

And **OT**, why do you not upgrade this to Win10? That is the path I took. This Win10 64-bit machine was formerly Win7 64-bit... and I was offered a **free** upgrade... but that is another topic...

 2. Server 2003 - 32-bit

Well, until you **upgrade** this to a 64-bit OS, you can only run **32-bit Tidy 5.4**. Of course, for **5.4 Tidy 32-bit**, you may have to also install the `VC 2015 Runtime 32 (x86)`...

Are you saying this does **not** work? Is this where you get `not a valid Win32 application`?

You advise **32-bit Tidy 5.2** is working. But so should **32-bit Tidy 5.4**, with the appropriate `runtimes`...

And am still **puzzled** why `tidy-5.5.10-win32-mt.zip` did not work! IIRC it is 32-bit, and does **not** need the VC runtimes... so should run in `Server 2003 - 32-bit`! What error message?

**Tidy Releases**

Concerning the next Tidy **Release**, that will be version 5.6... only even numbered minor versions are official releases... That is presently scheduled for mid August 2017, but if the past is anything to go on, that date may slide a few times, sometimes by months...

But we do keep the development version, `5.5`, presently `5.5.30`, in a very stable state, so it **can** be used in a production environment...

Just to check, with a few small patches, I **can** build Tidy `5.5.30` 64-bit, using MSVC10 (2010), but at this time have only tried the `64-bit` build... May get time later to build a 32-bit Tidy 5.5.30 using MSVC10, but that may be a few days...

So looking forward to further **clarification**... hopefully **success**... thanks...

**Comment by LWillms** [comment]
Yes, I have those two PCs, the server and the workstation. Also as third one a laptop which runs under Windows XP (because it has a MS Office 2007/2010 which I have inherited with the laptop itself, and would lose on replacing the OS). 

I ran the 32 bit Tidy on the 64 bit Win 7 system only to test if the problem is specific to the setup to my server. And yes, I installed both the 64 bit and the 32 bit VC 2015 runtime. 

On the two XP-level 32 bit PCs, I have installed only the 32 bit VC 2015 runtime, not the 64 bit version. 

On both of these systems, I get "this is no valid Win32 program" (or somesuch in German) when I try the 32 bit Tidy 5.4 on them, also with the MT build of Tidy 5.5 which I downloaded from the link in issue #515 

But this is no big problem for me, as far as I can see. Tidy 5.2 works fine, and the new line wrap / indentation coming in the next version are very nice to have, but not absolutely necessary. 

But the problem has to be solved. I'm ready to help with testing.

**Comment by geoffmcl** [comment]
@LWillms sorry to hear no complete success yet...

But yes, I really hope we can **solve** the problem, and **thanks** for your offer to help testing...

It turned out easier than I thought, to generate a current 32-bit Tidy 5.5.31, using MSVC10 (2010), so will try to embed a zip here...

Zip: 
[tidy-5.5.31-w32-vc10.zip](https://github.com/htacg/tidy-html5/files/1039208/tidy-5.5.31-w32-vc10.zip)

As the zip name implies, this is a 32-bit binary, built using MSVC10 (2010), the **same** as Tidy `5.2`... so if `5.2` runs, so should this - **I hope!**..

If you get the time, please test and advise... thanks...

**Comment by LWillms** [comment]
Success! That version built with VC10 functions on Windows with the XP code base, i.e. Windows XP and Windows Server 2003. 

My guess: either stick with VC 2010 or give up support for Windows XP, Windows Server 2003 and older operating systems.

**Comment by LWillms** [comment]
So, the problem is not, methinks, 32 bit Tidy vs. 64 bit Tidy, but Tidy on Windows XP and its server sister Server 2003 vs. on Windows 7 and its server sister Server 2008 R2, and later Windows versions. A test with a 64 bit Windows XP or Server 2003 should prove that (I do not have the means to test that). Such a test would probably show that the 64 bit Tidy 5.4 released in March would not run on a 64 bit Windows XP or 64 bit Windows Server 2003. And vice versa a 32 bit Tidy 5.4 as the one released in March would run on a 32 bit Windows 7 without any problem. 



(Server 2008 without an R corresponds to Windows Vista, an OS which I have avoided).

**Comment by LWillms** [comment]
I tested this Tidy 5.5.31-w32-vc10 with actual HTML files, and decided to keep it as my working Tidy. 

BTW, I tried the new vertical-space/wrap/indent config parameter as discussed in #486 , but these seem not to work yet. I thought they were scheduled for 5.5.

**Comment by geoffmcl** [comment]
@LWillms really thanks for testing tidy-5.5.31-w32-vc10.zip, and reporting... and glad it seems to work across the board...

While I agree somewhat with it is **not** simply `32 bit Tidy` vs. `64 bit Tidy`...

But for sure you can **not** run an app using 64-bit memory addressing in a 32-bit OS... Even though the particular CPU may have 64-bit registers, there is **no way** for the hardware drivers, etc to handle it... they have been built to **only** handle 32-bit addressing...

And while a 64-bit OS continues to support compatible modes, WOW, etc, it should always be able to run a 32-bit app... not sure how long MS will continue this support...

There does seem to be a **particular** difference between an app built with MSVC10 and MSVC14. 

First there is the difference between the required MSVC runtime `vcruntimeXXX.dll`... that is `100` for VC10, which almost seemed to be included in most OSES, and `140`, and some others, for VC14, which usually require the appropriate `vc_redist.xXX.exe` to also be installed...

**BUT** the use of **static** runtimes should overcome that! That is built with /MT vs. /MD...

I have again built a 32-bit Tidy, using **static** runtimes... it is in http://geoffair.org/tmp/tidy-5.5.31-w32-vc14-mt.zip. This 32-bit Tidy runs fine in my 2 32-bit Vista machines, and tomorrow, will try it in my 32-bit XP machine... I expect it to work...

But maybe you are **right**, that something **drastically** changed between XP, and later OSES...

It would be appreciated if you could get the time to download this, and give it a try... thanks...

While I know MS has given up on XP support, at this time I see no reason for Tidy to do this, while it costs very little extra effort... I talk about this in #550... and have recently added some small fixes to allow continued building with MSVC10... but would prefer using MSVC14, or later, hence the above zip... thanks...

**Comment by LWillms** [comment]
This your latest build does also not run on Server 2003. I didn't try the laptop with XP. 

I don't know in how far XP and 7 differ, but MS does not want to produce any more fixes for XP, and they have obviously removed the support for XP from their newer software development systems. 

BTW, Windows XP is actually Windows 4.0, and Vista 6.0 and Win 7 is 6.1. The big change is from XP to Vista, but they had first to iron out the teething problems in Vista, only the 6.1 i.e. Win7 is usable... 

"Don't change a running system" is a rule I learned early in my computing career, but there are applications which force me to move up:  Ultraedit 24 which finally has full support for UTF-8 does not run on XP-level OSes. And my current banking program does not support the other encryption method to which one of my banks switched to some time ago, and the new version does not work on XP 

Another fundamental rule of computing which I learned long time ago is to select the OS and the hardware according to the application one wants to use, not the other way 'round.

**Comment by LWillms** [comment]
Regarding VC15 runtime -- I just tried to install the 64 bit on my 32 bit Server 2003, and that was rejected with an error message saying that this can only be installed on Windows XP SP1 (X64) and newer platforms. 

And yes, I could install the 32 bit version of that runtime without problems. That lets me think that software build with VC2015 should be able to be executed on those systems where the runtime can bei installed. Otherwise it would not make sense to allow its installion on XP-level OSes.

**Comment by LWillms** [comment]
This entry on Stackoverflow shows the how to with screenshots, and some additional hints: 
https://stackoverflow.com/questions/35664861/how-to-target-windows-xp-in-microsoft-visual-studio-c

I found all this with a web search for "Microsoft Visual Studio C++ building software for XP". Auntie Google is your friend...

**Comment by geoffmcl** [comment]
@LWillms really thanks for following this and reporting...

I tried the `tidy-5.5.31-w32-vc14-mt.zip` in my **XP** machine, and as you reported got the `invalid win32` dialog, like in your Server 2003... BAH!

But you seem to have hit the nail on the head with the [stackoverflow](https://stackoverflow.com/questions/35664861/how-to-target-windows-xp-in-microsoft-visual-studio-c) link, and setting the `Platform Toolset` in MSVC14...

And while I would probably not characterize `Auntie Google is your friend`, it is certainly **very helpful** at times... ;=))

Thankfully this `Platform Toolset` can be done via cmake - just add `-T v140_xp` to the cmake config...

So the first I built was using the `default` **/MD** libraries, and while this EXE runs in XP, of course it complains about missing `VCRUNTIME140.dll`, and probably others, all of which I could probably be install... by the appropriate `vc_redist.x86.exe`... 

But I did **not** want to do that, in that venerable **old** machine... like you, it was the last machine where I had `Microsoft Office`, `Frontpage`, `Outlook`, `etc $$$$` installed, but have managed to wean myself of these `expensive` tools...

So then I built a `tidy 5.5.31 w32-vc14-mt-xp`, ie using the `static **/MT** libraries, and it ran fine in **XP**, and no doubt would also run in **Server 2003**... 

If you want to try it at this time I have only uploaded a http://geoffair.org/tmp/temp2.zip containing just the EXE... running it with `-v` should yield -

```
HTML Tidy for Widnows version 5.5.31.w32-vc14-mt-xp
```

So now it seems we have at least two(2) ways to continue supporting **XP**... use MSVC10, **or** MSVC14 using the `-T v140_xp` toolset... and this can probably be extended to later MSVC versions...

I hope you get the chance to try and report success with `temp2.zip`... thanks...

And yes, as you reported, there is no problem using this EXE, or others I sent, in a production environment... You do get all the advantages of the later version...

So maybe this issue has been **solved**, and can be closed... thanks...

**Comment by LWillms** [comment]
Success! 

tidy -v produces this: 

HTML Tidy for Windows version 5.5.31.w32-vc14-mt-xp

**Comment by LWillms** [comment]
Sorry for re-opening this issue, but only for a moment. I just want to mention that the Windows binaries downloadable from http://binaries.html-tidy.org/ should be replaced by versions compiled with the setup to make them runnable on Windows XP

**Comment by geoffmcl** [comment]
@LWillms you can see I have opened a #568 to remind myself what needs to be done...

My idea is to **not** replace, but **add** to the distribution...

PS: No need to re-open an issue to add a comment... we get them **all** anyway ;=))

**Comment by LWillms** [comment]
> tidy -v produces this:
> 
> HTML Tidy for Windows version 5.5.31.w32-vc14-mt-xp

Nice, but it does not produce any output. 

I go back to the vc10-version

**Comment by geoffmcl** [comment]
@LWillms thanks for your feedback... but no **idea** what you are saying here...

Of course you can **go back** to any version you prefer... no problem...

> Nice, but it does not produce any output. 

What does this mean? **totally** confused... in what system? what input? please explain... thanks...

**Comment by LWillms** [comment]
No output means no output. It simply terminated without anything. It just created the error file, when asked for with the -f option, but didn't write anything in it. All this on the same Windows Server 2003 which I do most of my work on. 

After rebooting the PC in the morning, Windows told me that TIDY.EXE had had a problem, and if I wanted to send a report to Microsoft. I prefer not to upload the dump etc here because it might contain information which I do not want to be publicly accessible, but I could mail the stuff to you privately. 

There were many instances of those error messages, but I retained only one.

**Comment by geoffmcl** [comment]
@LWillms I can only think you are running the **wrong** tidy.exe...

In my system I have up to **150** versions of `tidy.exe`, but **none** have that name! In fact in my system the command `tidy -v` will yield -

```
'tidy' is not recognized as an internal or external command,
operable program or batch file.
```

That is the way I deal with having **so** many versions, and **not** get confused... Each is given a unique name, usually related to their version. My oldest is `tidy2000.exe` -

```
<anywhere>tidy2000 -v
HTML Tidy release date: 4th August 2000
See http://www.w3.org/People/Raggett for details
```
Then the **latest** 

```
<anywhere>tidy5-5.5.31 -v
HTML Tidy for Windows version 5.5.31
```

You always need to **know** exactly which EXE is being run, if you have more than one... And that will depend on where you are at the time, as the system first searches locally, and then what is in one of the **PATH** directories... very easy to get it wrong...

So, in one sentence you say -
```
> tidy -v produces this:
> HTML Tidy for Windows version 5.5.31.w32-vc14-mt-xp
```

That is it produced an **output**...

Then, in the next sentence, after **rebooting**, tidy produced a **crash** dump, and no **output**... I can only ask **which** actual tidy EXE produced this? And what was the command line at the time...

I have no problem with you sending that **dump** directly to me, perhaps in a zip file. And maybe you could include the `tidy.exe` that caused it... this might help...

Hope we can get this sorted... thanks...

**Comment by LWillms** [comment]
You have no longer the exe which you produced 7 days ago announced in this message
https://github.com/htacg/tidy-html5/issues/566#issuecomment-305566823

> So then I built a tidy 5.5.31 w32-vc14-mt-xp, ie using the `static /MT libraries, and it ran fine in XP, and no doubt would also run in Server 2003...
> 
> If you want to try it at this time I have only uploaded a http://geoffair.org/tmp/temp2.zip containing just the EXE... running it with -v should yield -
> 
> HTML Tidy for Widnows version 5.5.31.w32-vc14-mt-xp
> 
> So now it seems we have at least two(2) ways to continue supporting XP... use MSVC10, or MSVC14 using the -T v140_xp toolset... and this can probably be extended to later MSVC versions...
> 
> I hope you get the chance to try and report success with temp2.zip... thanks...

Strange... 

Back then I unfortunately only checked if the program would start in the first place, just using the -v open. But trying to acutally generate an HTML output did not work. It would neither generated messages for the error file configured with the -f option. 

This Windows has the habit of informing the user about crashes from the previous session on booting into the next session, asks if information about it shoud be sent to Microsoft. This comes with a dialogue which I try to paste here, but as usual that does not work with Github. 

The text goes (in my translation to English) like this: "Tidy encountered an error and had to be terminated." 

And "Please report this error also to Microsoft" 

I would very much send the information gathered for informing microsoft, but that requires an eMail mailbox address.

**Comment by geoffmcl** [comment]
@LWillms this is getting very **frustrating**...

> Q: You have no longer the exe which you produced ... Strange...

**Of course I do**! It exists in **all** of my, Win 10 64-bit, Vista 32-bit, **and** XP 32-bit systems as -

```
 Directory of c:\mdos
2017-06-01  18:08  693,760 tidy-5.5.31.w32-vc14-mt-xp.exe
```

This `mdos` is just where I put **all** versions of `tidyXXXXX.exe`, and that directory, `C:\MDOS` is **permanently** in my `PATH` environment variable, so I can run **all** these versions from **anywhere**... especially against the formal [release](https://github.com/htacg/tidy-html5-tests) regression tests, and [my](https://github.com/geoffmcl/tidy-test/tree/master/test/input5) informal test files... hundreds and hundreds of them...

**AND** the - http://geoffair.org/tmp/temp2.zip - is still there... but that can be over-written anytime...

**AND** for me, this xp version runs fine in Win10, Vista and XP... so really unsure what you want me to do!!!

**LIKEWISE** I have a `tidy5-5.4.0-w32-mt-xp.exe` in my `MDOS`... this is from http://binaries.html-tidy.org/... `tidy-5.4.0-w32-mt-XP.zip` - downloaded, extracted, and renamed... and it **also** runs **fine**...

Sorry about the slightly **inconsistent** naming conventions I use... I am still **working this out**... that is what indicators to include in the naming convention... which I will apply to the `release` zips for Windows...

**SO** to me you have **two**, well 3++ if we count a vc10 versions, to use... Which one is giving you trouble? Are you using some other version, known to be a problem?

**What else do you want me to do?**

By clicking on my abundant pictures on GitHub, should put you in my profile page, where I do list a direct email address... or you could use `ubuntu _at_ geoffair _dot_ info` as the most active... or probably several other emails addresses I have... but essentially they **all** end up at `ubuntu`... my main linux Thunderbird email client...

But if you really **sort out** the version you are using, maybe you do **not** need to send me anything... and this can be **closed**... again... thanks...

**Comment by geoffmcl** [comment]
@LWillms ok, got your zip, **but** you did **not** include the `D:\UTIL\tidy.exe` as requested... please zip and send this... **exactly** this, no others...

From looking at the dumps, this appears to be the `tidy` from an earlier **`temp1.zip`**, since I can see MSVC14 dynamic DLLS there, like `api-ms-win-core-string-l1-1-0.dll` and others. When run I suspect this `tidy` would output `HTML Tidy for Windows version 5.5.31.w32-vc14-xp`, which is **not** exactly what you wrote above... **NOTE** the missing **-mt-**!!!

If you download and run the `tidy` from the **`temp2.zip`**, it will yield `HTML Tidy for Windows version 5.5.31.w32-vc14-mt-xp` - **NOTE** the added **-mt-**, that you wrote... **this is an important difference**

As suspected these crash **dump** do **not** give me much information... a copy of the actual `tidy.exe` would tell me more... analysis would give the exact date it was built, and more importantly exactly **which** system DLLS it would depend on... which is why I asked for a zip of the EXE...

Please **check**, and **re-check** the tidy version you are running... thanks...

**Comment by LWillms** [comment]
I just mailed the same archive back which I had downloaded from your site. I also included a screenshot from the command line window. 

T-Online did not want me to send it, rejected it with a 5.7.0 error message. I asked my ESET virus scanner to check the archive, and it could not find anything. Another server accepted it.

**Comment by geoffmcl** [comment]
@LWillms no **new** zip arrived? Even checked the spam folders... but I do **not** really need this... **forget** it...

I get it that you think you are mailing back what you downloaded, but the **dump** given indicates otherwise...

I will repeat **again** - the dump indicates `api-ms-win-core-string-l1-1-0.dll`, and others. 

The `tidy.exe` in `temp2.zip` does **not** use that DLL... look at the dumps yourself... the names are in unicode, but clearly visible... simple...

Or look at the EXE. I have my own [dump4](https://github.com/geoffmcl/dump4) utility, but open up a MSVC Command Prompt, and run `dumpbin /EXPORTS tidy.exe` and note the `Section contains the following imports:`... 

You should **only** find `KERNEL32.dll`... just **one** imported DLL in the **-mt-** version...

Then do the same with other tidy.exe versions you have... you will find, as well as `KERNEL32.dll`, `VCRUNTIME140.dll`, `api-ms-win-crt-stdio-l1-1-0.dll`, and others...

Either you sent the **wrong** dumps, or something... 

If you are truly running the **-mt-** version of tidy, which you seem to be **insisting**, ok, no problem... then at least send me the correct, matching **dump**... thanks...

**Comment by LWillms** [comment]
Well, it seems we are getting to nowhere with this. I manged to retain just one of the half a dozen or more error reports which Windows asked me to forward to Microsoft. When that doesn't fit the exe which doesn't produce output, I am at end. 

Anyway, as I wrote before, I am happy with the version which is built with VC10, and then want to leave it here. 

I have just one question -- it seems that the -v or -version command line option gets Tidy.exe to just display the version string, but not processing any given file. Is that right? Isn't there a way to get Tidy.exe to write the version into the err file, as I call that, i.e. the file specified by the -f command line option?

**Comment by LWillms** [comment]
HI, @Geoffmcl, your first line was cryptic to me and I did understand its meaning only earlier today, when I opened the mail account from which I tried to send the ZIP a second time for the first time since Friday evening, and more by accident found a non-delivery report where Google wrote that the content was suspicious to them. See your inbox.

**Comment by geoffmcl** [comment]
@LWillms no zip received, but as indicated I do **not** think this would help in **any** way... so please **forget** it...

Am thinking of the following **stern** warning -

**WARNING**: While some effort has been made to provide a Windows **`XP`** version of Tidy.exe, using MSVC14, this has **not** proved completely successful in **all** cases. We do **not** know exactly why!

**Regrettably**, if you have an **`XP`** like system, which seems to include `Windows Server 2003`, and this `Tidy.exe` fails to run correctly, for any reason, then you have no other options but to upgrade the system to a later version of Windows, and/or if you have the capability and tools, maybe build tidy from source in that system...

**Alternatively**, apparently some **older** versions of **`Tidy.exe`** may still run...

If you have this problem, then we can only state `HTML Tidy` no longer supports such **old** systems, and will **not accept any issue on the subject! Sorry**!

Does that make the position clear? Do you have any other suggestions about the warning? Over time I will find some places to **publish** this information, so users can be **warned**... thanks...

Concerning your question about `-v`, `-version`, even `--version`, when this command line item is reached, as you state, tidy will output the version string to stdout, and exit(0). No further processing will take place...

Now, if this `version` option is preceded by other commands, like say `-f temp1.txt`, then yes, this file might be created, but it may be blank... 

Tidy processes the command line in a completely **linear** way, so you can have like `-f temp1.txt first.html -v second.html`, where the `first.html` will be processed before it reaches the `-v`, but then, here the `second.html` would not be processed... and so on...

The tidy man page, and even `tidy -?` shows this ability to repeat options and files, on the command line, like -

```
tidy [options...] [file...] [options...] [file...]
```

And one thing that can still catch me now and then is when no **`file`** is given, like `tidy [options...]<enter>`!

Tidy will appear to **hang** since it will attempt to read the input from stdin... only `Ctrl + c` gets back to the prompt... so you can do like `echo para 1 | tidy [options...]` and get a html output including 'para 1'... Not something normally done in windows, but I am assured common in unix...

I have not particularly come across many other apps that will do all this... **`tidy`** may be **unique** ;=))

**Comment by LWillms** [comment]
As I said earlier, I am happy with the Tidy which you built with Visual Studio 2010, which identifies itself as "Tidy 5.5.31.w32-vc10"; this does work. I thought you would be interested in finding out why the one built with VC14 and this XP-option does not, but so be it. I have a working environment, and hope to soon upgrade the server to 2008 R2. Cheers... 
Thanks for all your attention!

**Comment by LWillms** [comment]
BTW2, the reason why the VC14-xp did not produce any output on my system might be that the TEMP2.ZIP contained only the EXE, but not the corresponding DLL.

**Comment by LWillms** [comment]
BTW2, the reason why the VC14-xp did not produce any output on my system might be that the TEMP2.ZIP contained only the EXE, but not the corresponding DLL.

**Comment by geoffmcl** [comment]
@LWillms yes, you had already informed us that `Tidy 5.5.31.w32-vc10` worked... thanks... but, as stated, I was hoping to **drop** the use of that 2010 tool, and use only vc14 2015, or later... may have to rethink that... at least for the next release...

And while I **would** be interested in finding out why the one built with vc14 failed in **some** environments, and that `tidy.exe`, as indicated, only has a dependence on `KERNEL32.DLL`, no others, so no need for `tidy.dll` - it runs, works fine, in my **XP** installation - it seems we have spent enough time chasing this...

OT: BTW, I am well aware of the `queue of filters`, and thanks for the reminder. That was not the **unique** thing I was referring to...

While there are many programs that support `-options file1 file2 file3 ...` - that is quite normal... and is supported by the GNU `getopt` library, putting all the non-options at the end of the list, so does even support out-of-order `-options`...

But Tidy's pure **linear** support for `-option1 file1 <process> -option2 file2 <process> ...` is **not** common, in my experience... although I am sure there **are** others... and do not need a list on this OT subject...

Accordingly, am marking this `Won't` fix, which in this case means more like, tried **many** things, spent a lot of time on it, and **can't** fix... and have some ideas for the future... thanks for all the feedback...

Can we now close this overly long saga? I have supplied you with a **working** up-to-date version `5.5.31.w32-vc10`... What more is there? Thanks...
